tao_compiler/mlir/disc/tools/disc-transform/TransformOps/TransformOpsExt.cc

```diff
@@ -191,6 +194,312 @@ DiagnosedSilenceableFailure DISCBufferizeOp::apply(
+        newOffsets.push_back(std::get<1>(z));
+      } else if (intRhsAttr && intRhsAttr.getInt() == 0) {
+        newOffsets.push_back(std::get<0>(z));
+      } else {
+        return failure();
+      }
+    }
+
+    rewriter.replaceOpWithNewOp<tensor::ExtractSliceOp>(
+        op, producerOp.getSource(), newOffsets, op.getMixedSizes(),
+        op.getMixedStrides());
+    return success();
+  }
+};
+
+/// Fold a tensor.extract_slice op if we can know that it's a full select slice.
+///
+/// convert:
+///   %0 = tensor.extract_slice %arg0[%o0, %o1][%s0, %s1][1, 1]
+///   %1 = destination_style_op(%0)
+///   %2 = tensor.extract_slice %1[0, 0][%s0, %s1][1, 1] // full select
+///   use(%2)
+/// to:
+///   %0 = tensor.extract_slice %arg0[%o0, %o1][%s0, %s1][1, 1]
+///   %1 = destination_style_op(%0)
+///   use(%1)
+struct FoldFullSelectedExtractSlicePattern
+    : public OpRewritePattern<tensor::ExtractSliceOp> {
+  using OpRewritePattern<tensor::ExtractSliceOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::ExtractSliceOp op,
+                                PatternRewriter& rewriter) const override {
+    if (!offsetAllZeros(op) || !strideAllOnes(op)) return failure();
+
+    auto sliceOp = getSliceProducer(op.getSource());
+    if (!sliceOp || !hasSameShape(op, sliceOp)) return failure();
+    rewriter.replaceOp(op, op.getSource());
+    return success();
+  }
+};
+
+/// Fold a self assigned tensor.insert_slice op.
+///
+/// convert:
+///   %0 = tensor.extract_slice %arg0[%o0, %o1][%s0, %s1][1, 1]
+///   %1 = destination_style_op(%0)
+///   %2 = vector.transfer_write %vec, %1[%c0, %c0]
+///   %3 = tensor.insert_slice %2 into %1[0, 0] [%s0, %s1] [1, 1]
+///   use(%3)
+/// to:
+///   %0 = tensor.extract_slice %arg0[%o0, %o1][%s0, %s1][1, 1]
+///   %1 = destination_style_op(%0)
+///   %2 = vector.transfer_write %vec, %1[%c0, %c0]
+///   use(%2)
+struct FoldSelfInsertSlicePattern
+    : public OpRewritePattern<tensor::InsertSliceOp> {
+  using OpRewritePattern<tensor::InsertSliceOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::InsertSliceOp op,
+                                PatternRewriter& rewriter) const override {
+    if (!offsetAllZeros(op) || !strideAllOnes(op)) return failure();
+
+    auto sourceSliceOp = getSliceProducer(op.getSource());
+    auto destSliceOp = getSliceProducer(op.getDest());
+    if (!sourceSliceOp || !destSliceOp) return failure();
+    // 1, dest & source have the same shape
+    if (!hasSameShape(sourceSliceOp, destSliceOp)) return failure();
+    // 2, the inserted tile from source and the dest have the same shape.
+    if (!hasSameShape(sourceSliceOp.getMixedSizes(), op.getMixedSizes()))
+      return failure();
+
+    rewriter.replaceOp(op, op.getSource());
+    return success();
+  }
+};
+
+/// convert:
+///   %0 = tensor.extract_slice %arg0[0, 0] [%s0, %s1] [1, 1] : tensor<?x?xf32>
+///   to tensor<?x?xf32> %1 = linalg.fill ins(%cst : f32) outs(%0 :
```

Comment:
nit: malformed...

Reply:
ohhh, it's re-formatted by clang-format. I'll send a new PR today. To reduce CI overhead, I'll change it in that PR.